pkg/azure/provider/helpers/driver.go

```diff
@@ -0,0 +1,678 @@
+	templateArgs := make([]any, 0, 3)
+	// NOTE: preserve the same order as these are ordered parameters which will be used for substitution.
+	templateArgs = append(templateArgs, resourceGroup)
+	for k := range providerSpecTags {
+		if strings.HasPrefix(k, utils.ClusterTagPrefix) || strings.HasPrefix(k, utils.RoleTagPrefix) {
+			templateArgs = append(templateArgs, k)
+		}
+	}
+	return templateArgs
+}
+
+func createVMNameMapperFn(suffix *string) accesshelpers.MapperFn[string] {
+	return func(r map[string]interface{}) *string {
+		if resourceNameVal, keyFound := r["name"]; keyFound {
+			resourceName := resourceNameVal.(string)
+			if suffix != nil && strings.HasSuffix(resourceName, *suffix) {
+				return to.Ptr(resourceName[:len(resourceName)-len(*suffix)])
+			}
+			return to.Ptr(resourceName)
+		}
+		return nil
+	}
+}
+
+// Helper functions used for driver.DeleteMachine
+// ---------------------------------------------------------------------------------------------------------------------
+
+// SkipDeleteMachine checks if ResourceGroup exists. If it does not exist then there is no need to delete any resource as it is assumed that none would exist.
+func SkipDeleteMachine(ctx context.Context, factory access.Factory, connectConfig access.ConnectConfig, resourceGroup string) (bool, error) {
+	resGroupAccess, err := factory.GetResourceGroupsAccess(connectConfig)
+	if err != nil {
+		return false, status.Error(codes.Internal, fmt.Sprintf("failed to create ResourceGroup access to process request: [ResourceGroup: %s]", resourceGroup))
+	}
+	resGroupExists, err := accesshelpers.ResourceGroupExists(ctx, resGroupAccess, resourceGroup)
+	if err != nil {
+		return false, status.Error(codes.Internal, fmt.Sprintf("failed to check if ResourceGroup %s exists, Err: %v", resourceGroup, err))
+	}
+	return !resGroupExists, nil
+}
+
+// GetDiskNames creates disk names for all configured OSDisk and DataDisk in the provider spec.
+func GetDiskNames(providerSpec api.AzureProviderSpec, vmName string) []string {
+	dataDisks := providerSpec.Properties.StorageProfile.DataDisks
+	diskNames := make([]string, 0, len(dataDisks)+1)
+	diskNames = append(diskNames, utils.CreateOSDiskName(vmName))
+	if !utils.IsSliceNilOrEmpty(dataDisks) {
+		for _, disk := range dataDisks {
+			diskName := utils.CreateDataDiskName(vmName, disk)
+			diskNames = append(diskNames, diskName)
+		}
+	}
+	return diskNames
+}
+
+// CheckAndDeleteLeftoverNICsAndDisks creates tasks for NIC and DISK deletion and runs them concurrently. It waits for them to complete and then returns a consolidated error if there is any.
+// This method will be called when these resources are left without an associated VM.
+func CheckAndDeleteLeftoverNICsAndDisks(ctx context.Context, factory access.Factory, vmName string, connectConfig access.ConnectConfig, providerSpec api.AzureProviderSpec) error {
```

Comment:
the function says it checks for Nic/disks ... but as per the flow it directly deletes them and returns error.. this way could we get stuck on a notFound error?

Reply:
This should not be an issue. The new Azure SDK API that deletes a NIC/Disk will not return an error if an attempt is made to delete a resource that does not exist. It is considered a successful delete, essentially it is treated as a NO-OP.

Reply:
could you add a comment to mention this

Reply:
Done.